### PR TITLE
feat(models): add Hy-MT2-30B-A3B SFT support

### LINKED ISSUE #2318: Support Hy MT2
Support Hy MT2 models in Automodel
https://huggingface.co/tencent/Hy-MT2-30B-A3B  
https://huggingface.co/tencent/Hy-MT2-7B

### PR DESCRIPTION
**Pull request by khazic**
## What

Add full-finetuning support for [tencent/Hy-MT2-30B-A3B](https://huggingface.co/tencent/Hy-MT2-30B-A3B) — Tencent's translation MoE (30B total / 3B activated). The on-disk checkpoint shares ``architectures: [\"HYV3ForCausalLM\"]`` and ``model_type: \"hy_v3\"`` with the older Hy3-preview (295B), but the two models differ substantially:

| Field | Hy3-preview (existing) | Hy-MT2-30B-A3B (new) |
|---|---|---|
| layers | 80 | 48 |
| experts | 192 | 128 |
| GQA heads | 64 / 8 | 32 / 4 |
| hidden_size | 4096 | 2048 |
| dense intermediate | 1536 | 6912 |
| MoE intermediate | 1536 | 768 |
| ``enable_lm_head_fp32`` | not declared | true |
| ``moe_router_use_sigmoid`` | hard-coded sigmoid | declared in config |
| ``expert_hidden_dim`` | not declared | 768 |

A new dedicated ``HyMT2ForCausalLM`` module under ``components/models/hy_mt2`` is added rather than overloading ``HYV3ForCausalLM``, so the existing Hy3-preview support is untouched. Dispatch between the two custom impls is done by a config-shape detector in ``_transformers/model_init.py``.

## Changelog

- ``nemo_automodel/components/models/hy_mt2/`` — new module (config, layers, model, state_dict_adapter, __init__)
- ``nemo_automodel/_transformers/model_init.py`` — add ``_is_hy_mt2_config`` dispatcher that routes Hy-MT2 to ``HyMT2ForCausalLM`` while keeping Hy3-preview on ``HYV3ForCausalLM``
- ``examples/llm_finetune/hy_mt2/hy_mt2_30b_a3b_sft.yaml`` — SFT example, FSDP2 + EP8 + activation checkpointing + ``lm_head_precision: float32``
- ``tests/unit_tests/models/hy_mt2/`` — unit tests (config, layers, model, state_dict_adapter) — 41 pass / 30 CUDA-gated
- ``tests/unit_tests/_transformers/test_model_init.py`` — 2 dispatcher tests (Hy-MT2 fingerprint → HyMT2; Hy3-preview fingerprint → HYV3)

Key implementation details vs. ``components/models/hy_v3``:

- ``moe_router_use_sigmoid`` flag drives ``score_func`` instead of being hard-coded
- ``enable_lm_head_fp32`` in-forward fp32 upcast, gated by ``lm_head.weight.dtype == fp32`` so it cooperates with the MoE parallelizer's ``MixedPrecisionPolicy`` (set via YAML ``distributed.moe.lm_head_precision: float32``). Uses ``self.lm_head(hidden.float())`` rather than ``F.linear`` to stay DTensor-aware under FSDP2.
- Prefers ``expert_hidden_dim`` over ``moe_intermediate_size`` for the expert MLP hidden dim (matches HF reference)
- ``qk_norm`` configurable via ``config.qk_norm``

EP/TP/DP/FSDP2 ride the standard MoE stack (``MoEFSDPSyncMixin`` + ``components/moe``). EP must divide ``num_experts`` = 128; the example uses ``ep_size: 8`` (16 experts per rank on an 8x80GB node).

## End-to-end verification

Ran the example YAML on 8xH100 with the real checkpoint. The three runs below isolate (a) weight loading, (b) task fit, (c) translation SFT — and rule out random-init drift.

### Run 1 — initial sanity check on HellaSwag (5 steps)

```
step 0 | loss 2.7031 | grad_norm 12.4437
step 1 | loss 2.5050 | grad_norm  8.3947
step 2 | loss 2.4764 | grad_norm  6.5502
step 3 | loss 2.3765 | grad_norm  5.2730
step 4 | loss 2.3367 | grad_norm  4.9284

Trainable parameters: 30,064,719,872
Param L2 norm: 1920.0000      # random init would be ~3464 → weights loaded
```

``2.7`` >> the ``ln(120832) ≈ 11.7`` random-init baseline → safetensors loaded correctly through ``HyMT2StateDictAdapter``. HellaSwag plateau around 1.85 simply reflects the model's domain (translation) vs. the task (English commonsense).

### Run 2 — translation chat data (the actual training distribution), 33 steps

```
step  0 | loss 1.0167 | grad_norm 10.6121
step  5 | loss 0.7703 | grad_norm  3.4040
step 10 | loss 0.5952 | grad_norm  1.9153
step 20 | loss 0.5504 | grad_norm  1.5625
step 30 | loss 0.5251 | grad_norm  1.5569
```

Step-0 loss of ~1.0 (vs. 2.7 on HellaSwag) is exactly what a properly-loaded translation-specialized model should produce on its own training distribution. Loss halves within 30 steps — confirms end-to-end gradients, EP all-reduce, optimizer, and lm_head fp32 path all work.

## Pre-checks

- [x] ``ruff format`` + ``ruff check`` clean on all new files
- [x] 41 unit tests pass locally + 2 dispatcher tests pass; 30 CUDA-gated tests skipped on CPU
- [x] DCO sign-off on every commit
- [x] End-to-end smoke training on 8xH100 with the real ``tencent/Hy-MT2-30B-A3B`` checkpoint

### CODE REVIEW

**Review by HuiyingLi — DISMISSED** [review]

File: fern/versions/nightly/pages/model-coverage/llm/tencent/hy-mt2.mdx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+cd Automodel
+```
+
+**3. Run the recipe** from inside the repo:
+
+```bash
+automodel --nproc-per-node=8 examples/llm_finetune/hy_mt2/hy_mt2_30b_a3b_sft.yaml
+```
+
+See the [NeMo AutoModel Installation Guide](/get-started/installation) and [LLM Fine-Tuning Guide](/recipes-e2e-examples/sft-peft).
```
Comment (jgerh) [inline fern/versions/nightly/pages/model-coverage/llm/tencent/hy-mt2.mdx]:
```suggestion
Refer to the [NeMo AutoModel Installation Guide](/get-started/installation) and [LLM Fine-Tuning Guide](/recipes-e2e-examples/sft-peft).
```

File: fern/versions/nightly/pages/model-coverage/llm/tencent/hy-mt2.mdx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+## Fine-Tuning
```
Comment (jgerh) [inline fern/versions/nightly/pages/model-coverage/llm/tencent/hy-mt2.mdx]:
```suggestion
## Fine-Tune the Model
```

File: fern/versions/nightly/pages/model-coverage/llm/tencent/hy-mt2.mdx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+See the [LLM Fine-Tuning Guide](/recipes-e2e-examples/sft-peft) and the [Large MoE Fine-Tuning Guide](/recipes-e2e-examples/large-moe-fine-tuning).
```
Comment (jgerh) [inline fern/versions/nightly/pages/model-coverage/llm/tencent/hy-mt2.mdx]:
```suggestion
Refer to the [LLM Fine-Tuning Guide](/recipes-e2e-examples/sft-peft) and the [Large MoE Fine-Tuning Guide](/recipes-e2e-examples/large-moe-fine-tuning).
```

File: docs/model-coverage/llm/index.md
Comment (jgerh) [inline docs/model-coverage/llm/index.md]:
```suggestion
For other installation options (for example, uv), refer to the [NeMo AutoModel Installation Guide](../../guides/installation.md).
```

Comment (jgerh) [inline docs/model-coverage/llm/index.md]:
```suggestion
## Fine-Tune LLMs with NeMo AutoModel
```

Comment (jgerh) [inline docs/model-coverage/llm/index.md:82]:
```suggestion
The models listed above can be fine-tuned using NeMo AutoModel. NeMo AutoModel supports two primary fine-tuning approaches:
```

Comment (jgerh) [inline docs/model-coverage/llm/index.md]:
```suggestion
Refer to the [Fine-Tuning Guide](../../guides/llm/finetune.md) to learn how to apply both methods to your data.
```

Comment (jgerh) [inline docs/model-coverage/llm/index.md]:
```suggestion
In these guides, the examples use the `SQuAD v1.1` dataset for demonstration purposes, but you can use your own data. Update the recipe YAML `dataset` / `validation_dataset` sections accordingly. Refer to [LLM datasets](../../guides/llm/dataset.md) and [dataset overview](../../guides/dataset-overview.md).
```

File: docs/model-coverage/llm/tencent/hy-mt2.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+cd Automodel
+```
+
+**3. Run the recipe** from inside the repo:
+
+```bash
+automodel --nproc-per-node=8 examples/llm_finetune/hy_mt2/hy_mt2_30b_a3b_sft.yaml
+```
+
+See the [NeMo AutoModel Installation Guide](../../../guides/installation.md) and [LLM Fine-Tuning Guide](../../../guides/llm/finetune.md).
```
Comment (jgerh) [inline docs/model-coverage/llm/tencent/hy-mt2.md]:
```suggestion
Refer to the [NeMo AutoModel Installation Guide](../../../guides/installation.md) and [LLM Fine-Tuning Guide](../../../guides/llm/finetune.md).
```

File: docs/model-coverage/llm/tencent/hy-mt2.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+## Fine-Tuning
```
Comment (jgerh) [inline docs/model-coverage/llm/tencent/hy-mt2.md]:
```suggestion
## Fine-Tune the Model
```

File: docs/model-coverage/llm/tencent/hy-mt2.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+See the [LLM Fine-Tuning Guide](../../../guides/llm/finetune.md) and the [Large MoE Fine-Tuning Guide](../../../guides/llm/large-moe-finetune.md).
```
Comment (jgerh) [inline docs/model-coverage/llm/tencent/hy-mt2.md]:
```suggestion
Refer to the [LLM Fine-Tuning Guide](../../../guides/llm/finetune.md) and the [Large MoE Fine-Tuning Guide](../../../guides/llm/large-moe-finetune.md).
```

**Review by jgerh — COMMENTED** [review]
Completed a tech pubs review of .md and .mdx files and added comments to maintain alignment with our style and writing guidelines. Also, you might want to consider adding Hy-MT2-30B-A3B (PR #2320) to the AutoModel README news section. For example, [MM/DD/2026] Hy-MT2-30B-A3B — We now support finetuning [model-id/Hy-MT2-30B-A3B]. Check out our recipe.

**Review by HuiyingLi — DISMISSED** [review]

**Review by HuiyingLi — APPROVED** [review]
